### ISSUE TITLE
OCR using tesseract

### ISSUE THREAD

**Issue by toxic0berliner** [sbs20/scanservjs#13]
It would be awesome to have a scanned file not only come out as a jpg (see enhancement request #12 )
But also have the option for pdf with ocr format.

I can help with that, I already have the one-liner bash that calls tesseract with a jpg input, and spits out a pdf file with the jpg in front and the OCR text behind the image ==> Very useful to be able to index documents since crawlers will see the OCR text, and humans when they open the pdf will still only see the fully correct jpg file in foreground...

**Comment by sbs20** [comment]
Yeah - it'd be a nice feature. I just want to make sure it will still work gracefully without.

Could you share the one-liner please? I'm being lazy, I know, but it'd help.

**Comment by toxic0berliner** [comment]
Sure, yere you go: 
`tesseract -l $ocrLanguage -psm 3 ${tmpFolder}/${name}.jpg ${tmpFolder}/${name} pdf`

for me it becomes: 
`tesseract -l fra -psm 3 /tmp/scan.jpg /tmp/scan pdf`

You just have to install the package `tesseract-ocr` along with `tesseract-ocr-eng` for english language, I add `tesseract-ocr-fra` for french (as I'm French...), I believe `tesseract-ocr-equ` and `tesseract-ocr-osd` come by default with tesseract, are usefull anyway for equations and other less-natural language scans.

**Comment by toxic0berliner** [comment]
Hmm, it just so happens that my tesseract version was quite old. Newer ones you'll find take `--psm` instead of `-psm` argument, but 3 is the default value anyway, so let me rephrase, the one-line is the most basic you'll need : 
`tesseract -l xxx file.jpg file pdf`
where `xxx` is the language code (eng, fra, ....
Beware, it's note file.pdf as output, but file[space] pdf, it's only the output basename, plus an extra parameter for the output format pdf, tesseract appends the extension by itself.

**Comment by xavery** [comment]
While introducing single-page OCR capabilities would certainly be welcome, the full functionality would need some way of creating multi-page documents and specifying the language(s) to use while making the conversion - these are obtainable from Tesseract itself by running `tesseract --list-langs`. Taking this to account, it seems like the most work is on the UI side : the only thing the server needs to do is run Tesseract with the chosen files and/or languages.

**Comment by toxic0berliner** [comment]
UI wise it's gonna be more tricky yes... Newer autonomous scanners have probably the simplest UX : after each page completes scanning, it asks yes/no is there more pages to continue scanning, if no it completed the document merging all successive pages into one off, that would also be the time to run the OCR. But language selection could be a global setting from my point of view, no need for a prompt by document.

**Comment by JnCrMx** [comment]
I have tried to do a small concept for scanning multiple pages (but with an option to enable it and a "finish" button) instead of a yes/no prompt.
It should be rather easy to use tesseract, simply by calling it instead of ``convert`` when merging the pages.

https://github.com/JnCrMx/scanservjs/tree/multiple-pages

**Comment by sbs20** [comment]
I haven't tested this yet, or had a chance to install / run.... but it's highly likely that this will work with the new pipelines configuration. If you add something like this to your pipelines. (it needs to read from stdin and write to stdout)

```json
    {
      extension: 'pdf',
      description: 'PDF (OCR Française)',
      commands: [
        'tesseract -l fra - - pdf'
      ]
    },
```

**Comment by toxic0berliner** [comment]
Took me a while but I finally updated to the latest scanservjs, and I have not been able to use the built-in pipeline for PDF with OCR because of a dumb error I got : 
`` `${Config.tesseract} -l ${Config.ocrLanguage} -c stream_filelist=true - - pdf > scan-0001.pdf` `` both config variables turn out to be undefined.
Even if I have this my my config.local : 
```
config.ocrLanguage = 'fra';
config.tesseract = '/usr/bin/tesseract'; 
```

I suspect the uppercase C for Config is the culpable, not sure thought...
I did make myself a custom pipeline that worked like a charm : 
```
{
        extension: 'pdf',
        description: 'OCR | PDF (JPG | High quality)',
        commands: [
          'convert @- -quality 92 tmp-%d.jpg && ls tmp-*.jpg',
          'tesseract -l fra -c stream_filelist=true - - pdf > scan-0001.pdf',
          'ls scan-*.*'
        ]
      }
```
